src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs

```diff
@@ -390,4 +479,35 @@ internal void InvalidateCache()
         }
     }
 
+    private void AddCachedExtensionToExtensionMap<T>(
+        IDictionary<string, ISet<string>> extensionMap,
+        string extensionType,
+        IList<T> extensions) where T : TestPluginInformation
+    {
+        if (extensions == null) return;
+
+        extensionMap.Add(extensionType, new HashSet<string>());
+        foreach (var extension in extensions)
+        {
+            extensionMap[extensionType].Add(extension.IdentifierData);
```

Comment:
Populate the hashset, first then add it. Rather than looking it up from the extension dictionary on every iteration. You can just select IdentifierData from extensions and use hashset object initializer, something like this:

new HashSet<string>{ extensions.Select(e => e.IdentifierData) };